core/chain_util.go

```diff
@@ -346,3 +349,34 @@ func GetBlockByHashOld(db ethdb.Database, hash common.Hash) *types.Block {
 	}
 	return (*types.Block)(&block)
 }
+
+// returns a formatted MIP mapped key.
+func mipmapKey(num, level uint64) []byte {
+	key := new(big.Int).SetUint64(num / level * level)
+
+	return append(mipmapPre, key.Bytes()...)
+}
+
+// WriteMapmapBloom writes each address included in the receipts' logs to the
+// MIP bloom bin.
+func WriteMipmapBloom(db ethdb.Database, number uint64, receipts types.Receipts) {
+	for _, level := range MIPMapLevels {
+		key := mipmapKey(number, level)
+		bloomDat, _ := db.Get(key)
+		bloom := types.BytesToBloom(bloomDat)
+		for _, receipt := range receipts {
+			for _, log := range receipt.Logs() {
+				bloom.Add(log.Address.Big())
+			}
+		}
+		db.Put(key, bloom.Bytes())
```

Comment:
I agree. I'll refactor the `PutTransaction` method from `transaction_util` and check for errors (and abort chain insertion).